Marlin/src/feature/bedlevel/bdl/bdl.cpp

```diff
@@ -42,151 +43,199 @@
 
 // M102 S-5   Read raw Calibrate data
 // M102 S-6   Start Calibrate
 // M102 S4    Set the adjustable Z height value (e.g., 'M102 S4' means it will do adjusting while the Z height <= 0.4mm , disable with 'M102 S0'.)
 // M102 S-1   Read sensor information
 
 #define MAX_BD_HEIGHT                 4.0f
+#define CMD_READ_VERSION              1016
 #define CMD_START_READ_CALIBRATE_DATA 1017
 #define CMD_END_READ_CALIBRATE_DATA   1018
 #define CMD_START_CALIBRATE           1019
 #define CMD_END_CALIBRATE             1021
-#define CMD_READ_VERSION  1016
-
-I2C_SegmentBED BD_I2C_SENSOR;
-
 #define BD_SENSOR_I2C_ADDR            0x3C
 
+I2C_SegmentBED BD_I2C_SENSOR;
+float BDS_Leveling::pos_zero_offset;
 int8_t BDS_Leveling::config_state;
-uint8_t BDS_Leveling::homing;
-
-void BDS_Leveling::echo_name() { SERIAL_ECHOPGM("Bed Distance Leveling"); }
 
 void BDS_Leveling::init(uint8_t _sda, uint8_t _scl, uint16_t delay_s) {
-  int ret = BD_I2C_SENSOR.i2c_init(_sda, _scl, BD_SENSOR_I2C_ADDR, delay_s);
-  if (ret != 1) SERIAL_ECHOLNPGM("BD_I2C_SENSOR Init Fail return code:", ret);
-  config_state = 0;
+  config_state = BDS_IDLE;
+  const int ret = BD_I2C_SENSOR.i2c_init(_sda, _scl, BD_SENSOR_I2C_ADDR, delay_s);
+  if (ret != 1) SERIAL_ECHOLNPGM("BD Sensor Init Fail (", ret, ")");
+  sync_plan_position();
+  pos_zero_offset = planner.get_axis_position_mm(Z_AXIS) - current_position.z;
+  SERIAL_ECHOLNPGM("BD Sensor Zero Offset:", pos_zero_offset);
+}
+
+bool BDS_Leveling::check(const uint16_t data, const bool hicheck/*=false*/) {
+  if (BD_I2C_SENSOR.BD_Check_OddEven(data) == 0)
+    SERIAL_ECHOLNPGM("Read Error.");
+  else if (!good_data(data))
+    SERIAL_ECHOLNPGM("Invalid data, please calibrate.");
+  else if (hicheck && (data & 0x3FF) > 550)
+    SERIAL_ECHOLNPGM("BD Sensor mounted too high!");
+  else if ((data & 0x3FF) >= (MAX_BD_HEIGHT) * 100 - 10)
+    SERIAL_ECHOLNPGM("Out of Range.");
+  else
+    return false;
+  return true; // error
+}
+
+float BDS_Leveling::interpret(const uint16_t data) {
+  return (data & 0x3FF) / 100.0f;
 }
 
 float BDS_Leveling::read() {
-  const uint16_t tmp = BD_I2C_SENSOR.BD_i2c_read();
-  float BD_z = NAN;
-  if (BD_I2C_SENSOR.BD_Check_OddEven(tmp) && (tmp & 0x3FF) < 1020)
-    BD_z = (tmp & 0x3FF) / 100.0f;
-  return BD_z;
+  const uint16_t data = BD_I2C_SENSOR.BD_i2c_read();
+  return check(data) ? NAN : interpret(data);
 }
 
 void BDS_Leveling::process() {
- //if (config_state == 0) return;
- static millis_t next_check_ms = 0; // starting at T=0
- static float z_pose = 0.0f;
- const millis_t ms = millis();
- if (ELAPSED(ms, next_check_ms)) { // timed out (or first run)
-    next_check_ms = ms + (config_state < 0 ? 1000 : 100);   // check at 1Hz or 10Hz
-
-    unsigned short tmp = 0;
-    const float cur_z = planner.get_axis_position_mm(Z_AXIS); //current_position.z
-    static float old_cur_z = cur_z,
-                 old_buf_z = current_position.z;
-
+  //if (config_state == BDS_IDLE) return;
```

Comment:
Considering that `bdl.process()` is called very frequently from `idle()` it seems like uncommenting this line would be a reasonable optimization. It's probably pretty costly to do i2c polling 5 times per second, and I see that if debugging (`DEBUG_OUT_BD`) is enabled this will also lead to an enormous amount of serial output. Also, setting the Status Message can be very costly and also lead to a lot of serial output and extra processing due to all the activity that `ui.set_status` can initiate.

Apparently this could also lead to `stepper.apply_directions` being called if there are any anomalies with the BD Sensor, and that might cause some mayhem in the middle of a long print job.

As mentioned in my other comment, we can probably remove the call to `stepper.apply_directions`, but in general anything we can do to reduce extra processing in `idle()` would be good.